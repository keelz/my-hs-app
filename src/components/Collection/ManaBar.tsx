import * as React from 'react';
import { IComponentProps } from '../../common/models/App.model';
import { composeClassname } from '../../common/utils';
import HSJSON from '../../common/constants/hsJson';
import { NavBar, NavItem } from '../NavBar';
import ManaGem from '../ManaGem';

interface IManaBarProps extends IComponentProps {}

export interface IManaBarStateProps {
  activeGem?: number;
  setFilter: (field: string, value: string) => any;
}

type Props = IManaBarProps & IManaBarStateProps;

export const composeManaGems = (): number[] =>
  Array(8).fill(null).map((_, i) => i);

export const composeClickAction = (action: (field: string, value: string) => any) =>
  (index: number) =>
  action(HSJSON.RESPONSE_PARAMS.COST, index.toString());

const ManaBar: React.SFC<Props> = props =>
  <NavBar
    id="collection-mana-bar"
    className={composeClassname('Collection-mana-bar shadow-md')(props.className)}>
    { composeManaGems().map(gem =>
      <NavItem
        key={`mana-bar-btn${gem}`}
        render={(_) => {
          return (
            <ManaGem
              className={props.activeGem === gem ? 'active' : ''}
              cost={gem}
              onClick={() => composeClickAction(props.setFilter)(gem)} />
          );
        }} />
    )}
  </NavBar>;

export default ManaBar;
