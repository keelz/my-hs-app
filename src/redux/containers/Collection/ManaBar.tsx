import { connect } from 'react-redux';
import HSJSON from '../../../common/constants/hsJson';
import ManaBar from '../../../components/Collection/ManaBar';
import { IRootState } from '../../Types';
import { collectionSetFilterAction } from '../../reducers/Collection';

const mapStateToProps = (state: IRootState) => ({
  activeGem: parseInt(state.Collection.filters[HSJSON.RESPONSE_PARAMS.COST], 10),
});

const mapDispatchToProps = {
  setFilter: collectionSetFilterAction,
};

const ConnectedManaBar = connect(mapStateToProps, mapDispatchToProps)(ManaBar);

export default ConnectedManaBar;
