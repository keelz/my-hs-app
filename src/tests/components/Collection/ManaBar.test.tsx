import * as React from 'react';
import * as enzyme from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import HSJSON from '../../../common/constants/hsJson';
import ManaBar, {
  composeLinkAction,
  composeManaGems
} from '../../../components/Collection/ManaBar';

const defaultProps = Object.freeze({
  setFilter: jest.fn(),
});

describe('Collection ManaBar', () => {
  describe('snapshots', () => {
    it('renders without crashing', () => {
      const tree = enzyme.shallow(<ManaBar {...defaultProps} />);
      expect(shallowToJson(tree)).toMatchSnapshot();
    });

    it('composes mana gems correctly', () => {
      const manaGems = composeManaGems();
      expect(manaGems).toMatchSnapshot();
    });
  });

  describe('helpers', () => {
    it('composes a link action correctly', () => {
      const testAction = jest.fn();
      const linkAction = composeLinkAction(testAction);
      linkAction(3);
      expect(testAction).toHaveBeenCalledWith(HSJSON.RESPONSE_PARAMS.COST, '3');
    });
  });
});
