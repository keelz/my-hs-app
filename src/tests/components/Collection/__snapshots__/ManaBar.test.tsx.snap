// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Collection ManaBar snapshots composes mana gems correctly 1`] = `
Array [
  0,
  1,
  2,
  3,
  4,
  5,
  6,
  7,
]
`;

exports[`Collection ManaBar snapshots renders without crashing 1`] = `
<div
  className="Collection-mana-bar"
>
  <Component
    id="collection-mana-bar"
  >
    <Component
      key="mana-bar-btn0"
      render={[Function]}
    />
    <Component
      key="mana-bar-btn1"
      render={[Function]}
    />
    <Component
      key="mana-bar-btn2"
      render={[Function]}
    />
    <Component
      key="mana-bar-btn3"
      render={[Function]}
    />
    <Component
      key="mana-bar-btn4"
      render={[Function]}
    />
    <Component
      key="mana-bar-btn5"
      render={[Function]}
    />
    <Component
      key="mana-bar-btn6"
      render={[Function]}
    />
    <Component
      key="mana-bar-btn7"
      render={[Function]}
    />
  </Component>
</div>
`;
